lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp

```diff
@@ -0,0 +1,394 @@
+    } else {
+      inlineInto(nestedPrefix, b, mapper, target);
+    }
+  }
+}
+
+void Inliner::inlineInstances(FModuleOp module) {
+  for (auto &op : llvm::make_early_inc_range(*module.getBodyBlock())) {
+    // If its not an instance op, skip it.
+    auto instance = dyn_cast<InstanceOp>(op);
+    if (!instance)
+      continue;
+
+    // If its not a regular module we can't inline it. Mark is as live.
+    auto module = symbolTable.lookup(instance.moduleName());
+    auto target = dyn_cast<FModuleOp>(module);
+    if (!target) {
+      liveModules.insert(module);
+      continue;
+    }
+
+    // If we aren't inlining the target, add it to the work list.
+    if (!shouldInline(target)) {
+      if (liveModules.insert(target).second) {
+        worklist.push_back(target);
+      }
+      continue;
+    }
+
+    // Create the wire mapping for results + ports. We RAUW the results instead
+    // of mapping them.
+    BlockAndValueMapping mapper;
+    OpBuilder b(instance);
+    auto instanceName = instance.nameAttr().getValue();
+    auto nestedPrefix = (instanceName + "_").str();
+    auto wires = mapPortsToWires(nestedPrefix, b, mapper, target);
+    for (unsigned i = 0; i < instance.getNumResults(); ++i)
+      instance.getResult(i).replaceAllUsesWith(wires[i]);
+
+    // Inline the module, it can be marked as flatten and inline.
+    if (shouldFlatten(target)) {
+      flattenInto(nestedPrefix, b, mapper, target);
+    } else {
+      inlineInto(nestedPrefix, b, mapper, target);
+    }
+
+    // Erase the replaced instance.
+    instance.erase();
+  }
+}
+
+Inliner::Inliner(CircuitOp circuit)
+    : circuit(circuit), context(circuit.getContext()), symbolTable(circuit) {
+  // Saving these attributes here makes lookup faster later.
+  auto inlineId = NamedAttribute(
+      Identifier::get("class", context),
+      StringAttr::get(context, "firrtl.passes.InlineAnnotation"));
+  inlineAttr = DictionaryAttr::getWithSorted(context, {inlineId});
+  auto flattenId = NamedAttribute(
+      Identifier::get("class", context),
+      StringAttr::get(context, "firrtl.transforms.FlattenAnnotation"));
+  flattenAttr = DictionaryAttr::getWithSorted(context, {flattenId});
+}
+
+void Inliner::run() {
+  // If the top module is not a regular module, there is nothing to do.
+  auto topModule = dyn_cast<FModuleOp>(symbolTable.lookup(circuit.name()));
+  if (topModule) {
+    worklist.push_back(topModule);
+  }
+  liveModules.insert(topModule);
+
+  // If the module is marked for flattening, flatten it. Otherwise, inline
+  // every instance marked to be inlined.
+  while (!worklist.empty()) {
+    auto module = worklist.pop_back_val();
+    if (shouldFlatten(module)) {
+      flattenInstances(module);
+      continue;
+    }
+    inlineInstances(module);
+  }
+
+  // Delete all unreferenced modules.
+  for (auto &op : llvm::make_early_inc_range(*circuit.getBody())) {
+    if (isa<FModuleOp, FExtModuleOp>(op) && !liveModules.count(&op)) {
```

Comment:
Circuit can only contain modules now that the terminator is gone, so you can delete the isa check (or change it into an assert)